src/eventbridge/eventpattern/index.ts

```diff
@@ -0,0 +1,798 @@
+      return handleBinary(expr);
+    } else if (isPropAccessExpr(expr)) {
+      return handlePropAccess(expr);
+    } else if (isUnaryExpr(expr)) {
+      return handleUnaryExpression(expr);
+    } else if (isCallExpr(expr)) {
+      return handleCall(expr);
+    } else {
+      throw new Error(`${expr.kind} is unsupported`);
+    }
+    // assertNever(expr);
+  };
+
+  const handleSmnt = (stmt: Stmt): PatternDocument => {
+    if (isReturn(stmt)) {
+      return handleExpr(stmt.expr);
+    } else {
+      throw new Error(`Unsupported statement ${stmt.kind}`);
+    }
+  };
+
+  const handleBinary = (expr: BinaryExpr): PatternDocument => {
+    if (expr.op === "==") {
+      return handleEquals(expr);
+    } else if (expr.op === "!=") {
+      return handleNotEquals(expr);
+    } else if ([">", ">=", "<", "<="].includes(expr.op)) {
+      return handleNumericRange(expr);
+    } else if (expr.op === "in") {
+      return handleInOperation(expr);
+    } else if (expr.op === "&&") {
+      return handleAnd(expr);
+    } else if (expr.op === "||") {
+      return handleOr(expr);
+    } else {
+      throw new Error(`Unsupported binary operator ${expr.op}`);
+    }
+  };
+
+  const handleAnd = (expr: BinaryExpr): PatternDocument => {
+    const left = handleExpr(expr.left);
+    const right = handleExpr(expr.right);
+
+    return andDocuments(left, right);
+  };
+
+  const andDocuments = (
+    classDocument1: PatternDocument,
+    classDocument2: PatternDocument
+  ): PatternDocument => {
+    const allKeys = [
+      ...new Set([
+        ...Object.keys(classDocument1.doc),
+        ...Object.keys(classDocument2.doc),
+      ]),
+    ];
+
+    return allKeys.reduce((doc, key) => {
+      // if key is only in one document, return it
+      const pattern = !(key in classDocument1.doc)
+        ? classDocument2.doc?.[key]
+        : !(key in classDocument2.doc)
+        ? classDocument1.doc?.[key]
+        : undefined;
+      // if the key is in both documents, try to merge them
+      const mergedPattern =
+        pattern ??
+        andMergePattern(key, classDocument1.doc[key], classDocument2.doc[key]);
+      return {
+        doc: {
+          ...doc.doc,
+          [key]: mergedPattern,
+        },
+      };
+    }, {} as PatternDocument);
+  };
+
+  const andMergePattern = (
+    key: string,
+    pattern1: PatternDocument | Pattern,
+    pattern2: PatternDocument | Pattern
+  ): PatternDocument | Pattern => {
+    // both are documents, lets merge the documents
+    if (isPatternDocument(pattern1) && isPatternDocument(pattern2)) {
+      return andDocuments(pattern1, pattern2);
+      // both are patterns, merge the patterns
+    } else if (!isPatternDocument(pattern1) && !isPatternDocument(pattern2)) {
+      if (isNumericAggregationPattern(pattern1)) {
+        if (isNumericAggregationPattern(pattern2)) {
+          return andNumericAggregation(pattern1, pattern2);
+        }
+        if (isNumericRangePattern(pattern2)) {
```

Comment:
nit: would prefer to use `else if` even when the branch returns - i think it groups related clauses together better than a separate `if` on a new line.

Reply:
Eventually I want to refactor these cases to support something like
1. If same type -> transform, error, continue
2. if either one is type -> transform, error, continue

But didn't want to start that yet...